library/database/class.database.php

```diff
@@ -149,30 +166,55 @@ public function commitTransaction() {
-        try {
-            $pDO = new PDO(strtolower($this->Engine).':'.$dsn, $user, $password, $this->ConnectionOptions);
-            $pDO->setAttribute(PDO::ATTR_EMULATE_PREPARES, 0);
-            $pDO->query("set time_zone = '+0:0'");
-
-            // We only throw exceptions during connect
-            $pDO->setAttribute(PDO::ATTR_ERRMODE, PDO::ERRMODE_SILENT);
-        } catch (Exception $ex) {
-            $timeout = false;
-            if ($ex->getCode() == '2002' && preg_match('/Operation timed out/i', $ex->getMessage())) {
-                $timeout = true;
-            }
-            if ($ex->getCode() == '2003' && preg_match("/Can't connect to MySQL/i", $ex->getMessage())) {
-                $timeout = true;
-            }
+        static $timeoutAt = null;
+        $timeoutAt = $timeoutAt ?? microtime(true) + ($this->SmoothTimeoutMillis / 1000);
+        do {
+            try {
+                $pDO = new PDO(strtolower($this->Engine).':'.$dsn, $user, $password, $this->ConnectionOptions);
+                $pDO->setAttribute(PDO::ATTR_EMULATE_PREPARES, 0);
+                $pDO->query("set time_zone = '+0:0'");
 
-            if ($timeout) {
-                throw new Exception(errorMessage('Timeout while connecting to the database', $this->ClassName, 'Connection', $ex->getMessage()), 504);
-            }
+                // We only throw exceptions during connect
+                $pDO->setAttribute(PDO::ATTR_ERRMODE, PDO::ERRMODE_SILENT);
 
-            throw new Exception(errorMessage('An error occurred while attempting to connect to the database', $this->ClassName, 'Connection', $ex->getMessage()), 500);
-        }
+                return $pDO;
 
-        return $pDO;
+            } catch (Exception $ex) {
+                if ($ex instanceof PDOException &&
+                    microtime(true) < $timeoutAt &&
+                    ($ex->getCode() === 1203 || $ex->getCode() === 1040 || $ex->getCode() === 1226)) {
+                    /*
+                     * https://dev.mysql.com/doc/refman/5.7/en/server-error-reference.html
+                     * ---
+                     * Error number: 1203; Symbol: ER_TOO_MANY_USER_CONNECTIONS; SQLSTATE: 42000
+                     * Message: User %s already has more than 'max_user_connections' active connections
+                     * ---
+                     * Error number: 1040; Symbol: ER_CON_COUNT_ERROR; SQLSTATE: 08004
+                     * Message: Too many connections
+                     * ---
+                     * Error number: 1226; Symbol: ER_USER_LIMIT_REACHED; SQLSTATE: 42000
+                     * Message: User '%s' has exceeded the '%s' resource (current value: %ld)
+                     * ---
+                     * In case connection limit is reached, we wait a random amount of time hoping the connection is released
+                     */
+                    usleep(rand($this->SmoothWaitMinMillis, $this->SmoothWaitMaxMillis) * 1000);
```

Comment:
Do you care whether this could result in a theoretical "worst" case where we wait for up to 5749ms vs your desired max of 5000?

Reply:
Your observation is quite good. I added a comment about it.
I wouldn't change the current behaviour as it would add more logic that doesn't add any extra value. All the setting are configurable to the desired values and this is a fallback mechanism that would be triggered only in rare cases